http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala

```diff
@@ -1,74 +1,280 @@
+  }
+  implicit val samProjectPolicyDecoder: Decoder[SamProjectPolicy] = Decoder.instance {
+    c =>
+      for {
+        policyName <- c.downField("accessPolicyName").as[AccessPolicyName]
+        project <- c.downField("resourceId").as[GoogleProject]
+      } yield SamProjectPolicy(policyName, project)
+  }
+  val subsystemStatusDecoder: Decoder[SubsystemStatus] = Decoder.instance {
+    c =>
+      for {
+        ok <- c.downField("ok").as[Boolean]
+        messages <- c.downField("messages").as[Option[List[String]]]
+      } yield SubsystemStatus(ok, messages)
+  }
+  implicit val systemsDecoder: Decoder[Map[Subsystem, SubsystemStatus]] = Decoder.decodeMap[Subsystem, SubsystemStatus](KeyDecoder.decodeKeyString.map(Subsystems.withName), subsystemStatusDecoder)
+  implicit val statusCheckResponseDecoder: Decoder[StatusCheckResponse] = Decoder.instance {
+    c =>
+      for {
+        ok <- c.downField("ok").as[Boolean]
+        systems <- c.downField("systems").as[Map[Subsystem, SubsystemStatus]]
+      } yield StatusCheckResponse(ok, systems)
+  }
+}
+
+final case class HttpSamDaoConfig(samUri: Uri,
+                                  petCacheEnabled: Boolean,
+                                  petCacheExpiryTime: FiniteDuration,
+                                  petCacheMaxSize: Int,
+                                  serviceAccountProviderConfig: ServiceAccountProviderConfig
+                                  )
+sealed trait AccessPolicyName extends Serializable with Product
+object AccessPolicyName {
+  final case object Creator extends AccessPolicyName {
+    override def toString = "creator"
   }
-}
\ No newline at end of file
+  final case object Owner extends AccessPolicyName {
+    override def toString = "owner"
+  }
+  final case class Other(asString: String) extends AccessPolicyName {
+    override def toString = asString
+  }
+
+  val stringToAccessPolicyName: Map[String, AccessPolicyName] = sealerate.collect[AccessPolicyName].map(p => (p.toString, p)).toMap
+
+}
+final case class SamNotebookClusterPolicy(accessPolicyName: AccessPolicyName, internalId: ClusterInternalId)
+final case class SamProjectPolicy(accessPolicyName: AccessPolicyName, googleProject: GoogleProject)
+final case class UserEmailAndProject(userEmail: WorkbenchEmail, googleProject: GoogleProject)
+
+sealed abstract class ResourceTypeName extends Product with Serializable {
+  type ResourcePolicy
+}
+object ResourceTypeName {
+  final case object NotebookCluster extends ResourceTypeName {
+    override def toString: String = "notebook-cluster"
+    override type ResourcePolicy = SamNotebookClusterPolicy
+  }
+  final case object BillingProject extends ResourceTypeName {
+    override def toString: String = "billing-project"
+    override type ResourcePolicy = SamProjectPolicy
+  }
+}
+
+final case object NotFoundException extends NoStackTrace
+final case class SamAuthProviderException(msg: String) extends LeoException(message = s"SamAuthProvider error: $msg")
```

Comment:
I think we landed on a friendlier error message in a separate PR. See https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/LeoAuthProviderHelper.scala#L19

Reply:
👍 ...I added traceId, but didn't add the `isTimeOut` boolean (with current setup, when max retry reaches, http4s will throw the same exception that caused it to fail. We'll see something like `[INFO] [22:42:39.401] [scala-execution-context-global-2185] org.http4s.client.middleware.Retry - Request threw an exception on attempt #4. Retrying after 4 milliseconds` in log, so hopeful it should be obvious enough. If not, we can revisit)